File: src/01_data_structures/03_queue/implementations/go/queue_test.go
package queue

import (
	"reflect"
	"testing"
)

func TestCreatesEmptyQueue(t *testing.T) {
	arrayQueue := ArrayQueue[int]{}

	if !reflect.DeepEqual(arrayQueue.isEmpty(), true) {
		t.Errorf("%v != %v", arrayQueue.isEmpty(), true)
	}
}

func TestEnqueueItemToEmptyQueue(t *testing.T) {
	arrayQueue := ArrayQueue[int]{}
	arrayQueue.enqueue(1)

	if !reflect.DeepEqual(arrayQueue.isEmpty(), false) {
		t.Errorf("%v != %v", arrayQueue.isEmpty(), false)
	}
}

func TestEnqueueTwoItemsToEmptyQueue(t *testing.T) {
	arrayQueue := ArrayQueue[int]{}
	arrayQueue.enqueue(1).enqueue(2)

	if !reflect.DeepEqual(arrayQueue.isEmpty(), false) {
		t.Errorf("%v != %v", arrayQueue.isEmpty(), false)
	}
}

func TestDequeueItemFromEmptyQueue(t *testing.T) {
	arrayQueue := ArrayQueue[int]{}
	dequeued, ok := arrayQueue.dequeue()

	if !reflect.DeepEqual(arrayQueue.isEmpty(), true) {
		t.Errorf("%v != %v", arrayQueue.isEmpty(), true)
	}

	if !reflect.DeepEqual(dequeued, 0) {
		t.Errorf("%v != %v", dequeued, 0)
	}

	if !reflect.DeepEqual(ok, false) {
		t.Errorf("%v != %v", ok, false)
	}
}

func TestDequeueItemFromQueue(t *testing.T) {
	arrayQueue := ArrayQueue[int]{}
	arrayQueue.enqueue(1).enqueue(2)

	firstItem, ok := arrayQueue.dequeue()
	if !reflect.DeepEqual(ok, true) {
		t.Errorf("%v != %v", ok, true)
	}
	if !reflect.DeepEqual(firstItem, 1) {
		t.Errorf("%v != %v", firstItem, 1)
	}

	secondItem, ok := arrayQueue.dequeue()
	if !reflect.DeepEqual(ok, true) {
		t.Errorf("%v != %v", ok, true)
	}
	if !reflect.DeepEqual(secondItem, 2) {
		t.Errorf("%v != %v", secondItem, 2)
	}
}

func TestReturnsLengthForEmptyQueue(t *testing.T) {
	arrayQueue := ArrayQueue[int]{}

	if !reflect.DeepEqual(arrayQueue.length(), 0) {
		t.Errorf("%v != %v", arrayQueue.length(), 0)
	}
}

func TestReturnsLengthForQueue(t *testing.T) {
	arrayQueue := ArrayQueue[int]{}
	arrayQueue.enqueue(1).enqueue(2)

	arrayQueue.dequeue()
	if !reflect.DeepEqual(arrayQueue.length(), 1) {
		t.Errorf("%v != %v", arrayQueue.length(), 1)
	}

	arrayQueue.dequeue()
	if !reflect.DeepEqual(arrayQueue.length(), 0) {
		t.Errorf("%v != %v", arrayQueue.length(), 0)
	}

	arrayQueue.dequeue()
	if !reflect.DeepEqual(arrayQueue.length(), 0) {
		t.Errorf("%v != %v", arrayQueue.length(), 0)
	}
}

func TestIsEmptyForEmptyQueue(t *testing.T) {
	arrayQueue := ArrayQueue[int]{}

	if !reflect.DeepEqual(arrayQueue.isEmpty(), true) {
		t.Errorf("%v != %v", arrayQueue.isEmpty(), true)
	}
}

func TestIsEmptyWhenQueueHasValues(t *testing.T) {
	arrayQueue := ArrayQueue[int]{}

	arrayQueue.enqueue(1)
	if !reflect.DeepEqual(arrayQueue.isEmpty(), false) {
		t.Errorf("%v != %v", arrayQueue.isEmpty(), false)
	}
}

func TestPeeks(t *testing.T) {
	arrayQueue := ArrayQueue[int]{}

	arrayQueue.enqueue(1).enqueue(2)
	firstItem := arrayQueue.peek()

	if !reflect.DeepEqual(firstItem, 1) {
		t.Errorf("%v != %v", firstItem, 1)
	}

	dequeuedFirstItem, ok := arrayQueue.dequeue()
	if !reflect.DeepEqual(ok, true) {
		t.Errorf("%v != %v", ok, true)
	}
	if !reflect.DeepEqual(dequeuedFirstItem, 1) {
		t.Errorf("%v != %v", dequeuedFirstItem, 1)
	}

	dequeuedSecondItem, ok := arrayQueue.dequeue()
	if !reflect.DeepEqual(ok, true) {
		t.Errorf("%v != %v", ok, true)
	}
	if !reflect.DeepEqual(dequeuedSecondItem, 2) {
		t.Errorf("%v != %v", dequeuedSecondItem, 2)
	}

	dequeuedThirdItem, ok := arrayQueue.dequeue()
	if !reflect.DeepEqual(ok, false) {
		t.Errorf("%v != %v", ok, false)
	}
	if !reflect.DeepEqual(dequeuedThirdItem, 0) {
		t.Errorf("%v != %v", dequeuedThirdItem, 0)
	}

	if !reflect.DeepEqual(arrayQueue.isEmpty(), true) {
		t.Errorf("%v != %v", arrayQueue.isEmpty(), true)
	}
}

File: src/01_data_structures/03_queue/implementations/rs/queue_test.rs
#[cfg(test)]
mod tests {
  use super::super::queue::ArrayQueue;

  #[test]
  fn creates_empty_queue() {
    let array_queue: ArrayQueue<i32> = ArrayQueue::new();

    assert_eq!(array_queue.is_empty(), true);
  }

  #[test]
  fn enqueue_item_to_empty_queue() {
    let array_queue: ArrayQueue<i32> = ArrayQueue::new().enqueue(1);

    assert_eq!(array_queue.is_empty(), false);
  }

  #[test]
  fn enqueue_two_items_to_empty_queue() {
    let array_queue: ArrayQueue<i32> = ArrayQueue::new().enqueue(1).enqueue(2);

    assert_eq!(array_queue.is_empty(), false);
  }

  #[test]
  fn dequeue_item_from_empty_queue() {
    let mut array_queue: ArrayQueue<i32> = ArrayQueue::new();
    let enqueued = array_queue.dequeue();

    assert_eq!(array_queue.is_empty(), true);
    assert_eq!(enqueued, None);
  }

  #[test]
  fn dequeue_item_from_queue() {
    let mut array_queue: ArrayQueue<i32> = ArrayQueue::new().enqueue(1).enqueue(2);

    let first_item = array_queue.dequeue();
    assert_eq!(first_item, Some(1));
    let second_item = array_queue.dequeue();
    assert_eq!(second_item, Some(2));
  }

  #[test]
  fn returns_length_for_empty_queue() {
    let array_queue: ArrayQueue<i32> = ArrayQueue::new();

    assert_eq!(array_queue.length(), 0);
  }

  #[test]
  fn returns_length_for_queue() {
    let mut array_queue: ArrayQueue<i32> = ArrayQueue::new().enqueue(1).enqueue(2);
    array_queue.dequeue();
    assert_eq!(array_queue.length(), 1);
    array_queue.dequeue();
    assert_eq!(array_queue.length(), 0);
    array_queue.dequeue();
    assert_eq!(array_queue.length(), 0);
  }

  #[test]
  fn is_empty_for_empty_queue() {
    let array_queue: ArrayQueue<i32> = ArrayQueue::new();
    assert_eq!(array_queue.is_empty(), true);
  }

  #[test]
  fn is_empty_when_queue_has_values() {
    let array_queue: ArrayQueue<i32> = ArrayQueue::new().enqueue(1);
    assert_eq!(array_queue.is_empty(), false);
  }

  #[test]
  fn peeks() {
    let mut array_queue: ArrayQueue<i32> = ArrayQueue::new().enqueue(1).enqueue(2);
    let first_item = array_queue.peek();
    assert_eq!(first_item, Some(1).as_ref());

    let dequeued_first_item = array_queue.dequeue();
    assert_eq!(dequeued_first_item, Some(1));
    let dequeued_second_item = array_queue.dequeue();
    assert_eq!(dequeued_second_item, Some(2));
    assert_eq!(array_queue.is_empty(), true);
  }
}
